### PR TITLE
Aura: Adds some compatibility mode to support old chains

### PR DESCRIPTION
**Pull request by bkchr**
In https://github.com/paritytech/substrate/pull/9132 we changed the way how we get the authorities from the runtime. Before this mentioned pr we would call `initialize_block` before calling the authorities runtime function. The problem with this was that when you have a block X that would switch the authority set, it would already be signed by an authority of the new set. This was wrong, as a block should only be signed by the current authority set. As this change is a hard fork, this pr brings back the possibility for users that have a chain running with this old logic to upgrade.

They will need to use:
```
CompatibilityMode::UseInitializeBlock { until: some_block_in_the_future }
```

Using this compatibility mode will make the node behave like the old nodes, aka calling `initialize_block` before doing the actual runtime call to `authorities`. Then when the given `until` block is being build/imported the node switches to the new behaviour of not calling `initialize_block` before. This is a hard fork, so the `until` block should be chosen wisely as a point where all nodes in the network have upgraded.

Closes: https://github.com/paritytech/substrate/issues/10103

cumulus companion: https://github.com/paritytech/cumulus/pull/1803

### CODE REVIEW

**Review by davxy — APPROVED** [review]
LGTM. Just a nit regarding the doc of `CompatibilityMode`

File: client/consensus/aura/src/lib.rs [hunk 1]
```diff
@@ -74,6 +74,41 @@ pub use sp_consensus_aura::{
...
+	///
+	/// By calling `initialize_block` before fetching the authorities, on a block that
+	/// would enact a new validator set, the block would already be build/sealed by an
+	/// authority of the new set. A block that enacts a new set, should not be sealed/build
+	/// by an authority of the new set. This isn't done anymore. However, to make new nodes
+	/// being able to sync the old chain this compatibility mode exists.
+	UseInitializeBlock {
+		/// The block number until this compatibility mode should be executed. The first runtime
+		/// call in the context (importing it/building it) of the `until` block should disable the
+		/// compatibility mode. This number should be of a block in the future! It should be a
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/consensus/aura/src/lib.rs:98]: I'm open to any improvement of this sentence. I already tried to make it more clear. :see_no_evil:

File: client/consensus/aura/src/lib.rs [hunk 2]
```diff
@@ -74,6 +74,41 @@ pub use sp_consensus_aura::{
...
+///
+/// This is required for when the chain was launched and later there
+/// was a consensus breaking change.
+#[derive(Debug, Clone)]
+pub enum CompatibilityMode<N> {
+	/// Don't use any compatibility mode.
+	None,
+	/// Call `initialize_block` before doing any runtime calls.
+	///
+	/// The node would execute `initialize_block` before fetchting the authorities
```
Comment (andresilva) [inline client/consensus/aura/src/lib.rs:87]:
```suggestion
	/// Previously the node would execute `initialize_block` before fetchting the authorities
```

File: client/consensus/aura/src/lib.rs [hunk 3]
```diff
@@ -74,6 +74,41 @@ pub use sp_consensus_aura::{
...
+	/// from the runtime. This behaviour changed in: <https://github.com/paritytech/substrate/pull/9132>
+	///
+	/// By calling `initialize_block` before fetching the authorities, on a block that
+	/// would enact a new validator set, the block would already be build/sealed by an
+	/// authority of the new set. A block that enacts a new set, should not be sealed/build
+	/// by an authority of the new set. This isn't done anymore. However, to make new nodes
+	/// being able to sync the old chain this compatibility mode exists.
```
Comment (andresilva) [inline client/consensus/aura/src/lib.rs:94]:
```suggestion
	/// authority of the new set. With this mode disabled (the default) a block that enacts a new set
	/// isn't sealed/built by an authority of the new set, however to make new nodes be able to sync
	/// old chains this compatibility mode exists.
```

File: client/consensus/aura/src/lib.rs [hunk 4]
```diff
@@ -74,6 +74,41 @@ pub use sp_consensus_aura::{
...
+	UseInitializeBlock {
+		/// The block number until this compatibility mode should be executed. The first runtime
+		/// call in the context (importing it/building it) of the `until` block should disable the
+		/// compatibility mode. This number should be of a block in the future! It should be a
+		/// block number on that all nodes have upgraded to a release that runs with the
+		/// compatibility mode. After this block there will be a hard fork when the authority set
+		/// changes, between the old nodes (running with `initialize_block`) and the new nodes.
```
Comment (andresilva) [inline client/consensus/aura/src/lib.rs:101]:
```suggestion
		/// The block number until this compatibility mode should be executed. The first runtime
		/// call in the context of the `until` block (importing it/building it) will disable the compatibility 
		/// mode (i.e. at `until` the default rules will apply). When enabling this compatibility mode the 
		/// `until` block should be a future block on which all nodes will have upgraded to a release that 
		/// includes the updated compatibility mode configuration. At `until` block there will be a hard 
		/// fork when the authority set changes, between the old nodes (running with `initialize_block`, 
		/// i.e. without the compatibility mode configuration) and the new nodes.
```

**Review by andresilva — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**